## PyTorch Review

* Tensor based deep learning library
* Accelerates computations on GPUs (able to switch between CPU and GPU)
* Automatic differentiation engine known as autograd paves the way of backpropagation and model optimization

Let's import torch and check if there is cuda available

In [3]:
import torch
torch.cuda.is_available()

False

### Tensors

Tensors generalizes vectors and matrices to higher dimensions.

* A scalar value (just a number) is a zero dimensional tensor.
* A vector is a one dimensional tensor.
* A matrix is a two dimensional tensor in PyTorch.

We can also define higher dimensional tensors like 3D tensor, and so forth which are efficiently computed in PyTorch.

Let's create some example PyTorch tensors.

In [4]:
tensor_0d = torch.tensor(1)  # zero dimensional tensor

tensor_1d = torch.tensor([1, 2, 3])  # 1D tensor

tensor_2d = torch.tensor([[1, 2],
                          [3, 4]])  # 2D tensor

tensor_3d = torch.tensor([[[1, 2], [3, 4]],
                          [[5, 6], [7, 8]]])  # 3D tensor

Let's check the shapes of the tensors defined above. We can use `.shape` attribute of the tensor to access it.

In [29]:
print("Is 0D tensor? : ", tensor_0d.shape)
print("Is 1D tensor? : ", tensor_1d.shape)
print("Is 2D tensor? : ", tensor_2d.shape)
print("Is 3D tensor? : ", tensor_3d.shape)

Is 0D tensor? :  torch.Size([])
Is 1D tensor? :  torch.Size([3])
Is 2D tensor? :  torch.Size([2, 2])
Is 3D tensor? :  torch.Size([2, 2, 2])


We can use `.reshape` or `.view` (the more common command for reshaping tensors in PyTorch) to reshape the tensor to a new dimension:

.view() requires the original data to be contiguous and will fail if it isn’t,
whereas .reshape() will work regardless, copying the data if necessary to ensure the desired shape.

What does "-1" do?

The "-1" in tensor reshaping is a placeholder that automatically infers the correct dimension size based on the total number of elements in the tensor and the other specified dimensions.

So, in `view((1, -1))`: 1 row and as many columns as needed to maintain the same number of elements which is 4 for this example.

<div class="alert alert-block alert-warning">
Also, pay attention the difference between a 1D tensor and 2D tensor with the same elements.
</div>

In [35]:
print("Reshape to a 2D tensor:")
print("-----------------------")
new_shape_tensor = tensor_2d.view((1, -1))

print(new_shape_tensor)
print("Shape: ", new_shape_tensor.shape, end="\n\n")

print("Reshape to a 1D tensor:")
print("-----------------------")
new_shape_tensor = tensor_2d.view((-1))

print(new_shape_tensor)
print("Shape: ", new_shape_tensor.shape, end="\n\n")

Reshape to a 2D tensor:
-----------------------
tensor([[1, 2, 3, 4]])
Shape:  torch.Size([1, 4])

Reshape to a 1D tensor:
-----------------------
tensor([1, 2, 3, 4])
Shape:  torch.Size([4])



#### Tensor Data Types

We can use `.dtype` attribute of a tensor to check the data type.

By default, PyTorch adopts the 64-bit integer data type from Python.

In [11]:
print("Data type of 0D tensor : ", tensor_0d.dtype)
print("Data type of 1D tensor : ", tensor_1d.dtype)
print("Data type of 2D tensor : ", tensor_2d.dtype)
print("Data type of 3D tensor : ", tensor_3d.dtype)

Data type of 0D tensor :  torch.int64
Data type of 1D tensor :  torch.int64
Data type of 2D tensor :  torch.int64
Data type of 3D tensor :  torch.int64


Let's also check the data type of float tensors.

By default, PyTorch adopts the 32-bit float data type from Python which is a really sufficient precision for most of the applications in deep learning while consuming less memory and computational resources than a 64-bit floating point number.

In [12]:
tensor_1d_float = torch.tensor([1.0, 2.0, 3.0])
print("Defined tensor: ", tensor_1d_float)
print("Type of tensor: ", tensor_1d_float.dtype)

Defined tensor:  tensor([1., 2., 3.])
Type of tensor:  torch.float32


We can change the type of the tensor by using `.to` method:

In [16]:
tensor_2d.to(torch.float32)

print("New data type: ", tensor_2d.to(torch.float32).dtype)

New data type:  torch.float32


#### Transpose of a Tensor

We can use .T to transpose a tensor, which means flipping it across its diagonal.

In [48]:
print("Transpose of a 2D tensor: ")
print("--------------------------")

print("Before transpose: ", tensor_2d.shape)
display(tensor_2d)

print("After transpose: ", tensor_2d.T.shape)
display(tensor_2d.T)

Transpose of a 2D tensor: 
--------------------------
Before transpose:  torch.Size([2, 2])


tensor([[1, 2],
        [3, 4]])

After transpose:  torch.Size([2, 2])


tensor([[1, 3],
        [2, 4]])

In [69]:
print("Transpose of a 3D tensor: ")
print("--------------------------")

tensor_3d = torch.tensor([[[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]],
                          [[13, 14, 15], [16, 17, 18], [19, 20, 21], [22, 23, 24]]])  # 3D tensor with the shape 2x4x3

print("Before transpose: ", tensor_3d.shape)
display(tensor_3d)

print("After transpose: ", tensor_3d.transpose(1, 2).shape)
display(tensor_3d.transpose(1, 2))


Transpose of a 3D tensor: 
--------------------------
Before transpose:  torch.Size([2, 4, 3])


tensor([[[ 1,  2,  3],
         [ 4,  5,  6],
         [ 7,  8,  9],
         [10, 11, 12]],

        [[13, 14, 15],
         [16, 17, 18],
         [19, 20, 21],
         [22, 23, 24]]])

After transpose:  torch.Size([2, 3, 4])


tensor([[[ 1,  4,  7, 10],
         [ 2,  5,  8, 11],
         [ 3,  6,  9, 12]],

        [[13, 16, 19, 22],
         [14, 17, 20, 23],
         [15, 18, 21, 24]]])

![01_PyTorch_transpose](../../images/01_PyTorch_transpose.png)

#### Matrix Multiplication

We can either use `.matmul` method or more practically `@` symbol for matrix multiplications.

In [78]:
print("Before multiplication:")
print(tensor_2d)
print("")

print("After multiplication using .matmul:")
print(tensor_2d.matmul(tensor_2d))
print("")

print("After multiplication using @:")
print(tensor_2d @ tensor_2d)

Before multiplication:
tensor([[1, 2],
        [3, 4]])

After multiplication using .matmul:
tensor([[ 7, 10],
        [15, 22]])

After multiplication using @:
tensor([[ 7, 10],
        [15, 22]])


### Computation Graph and Automatic Differentiation

PyTorch builds a computation graph in the background, and we can use this to calculate gradients to train the model and update the model parameters.

If we want to activate the backpropagation, we need to ensure that the `requires_grad` attribute must be set to True which is set by default.

PyTorch’s autograd engine constructs a computational graph in the background by tracking every operation performed on tensors. 
Then, calling the grad function, we can compute the gradient of the loss concerning the model parameters.

Below, we first calculate a basic arithmetic calculation as forward calculation. Then, we trigger the gradient calculation with respect to loss using `grad` method.

In [82]:
import torch.nn.functional as F
from torch.autograd import grad

y = torch.tensor([5.0])
x1 = torch.tensor([1.1])
w1 = torch.tensor([4.4], requires_grad=True)
b = torch.tensor([0.0], requires_grad=True)
z = x1 * w1 + b
print("z: ", z)

a = torch.relu(z)  # rectified linear unit (ReLU) activation function
print("a: ", a)

loss = F.mse_loss(a, y)  # mean squared error loss
print("loss: ", loss)

grad_L_w1 = grad(loss, w1, retain_graph=True)
grad_L_b = grad(loss, b, retain_graph=True)

print("grad_L_w1: ", grad_L_w1)
print("grad_L_b: ", grad_L_b)

z:  tensor([4.8400], grad_fn=<AddBackward0>)
a:  tensor([4.8400], grad_fn=<ReluBackward0>)
loss:  tensor(0.0256, grad_fn=<MseLossBackward0>)
grad_L_w1:  (tensor([-0.3520]),)
grad_L_b:  (tensor([-0.3200]),)


PyTorch will compute the gradients of all the leaf nodes in the graph using `backward` method, which will be stored via the tensors’ .grad attributes.

In [83]:
loss.backward()
print(w1.grad)
print(b.grad)

tensor([-0.3520])
tensor([-0.3200])


Handwritten gradient calculations can be seen in the following figure where the same values are obtained.

![02_Backward_calculation](../../images/02_Backward_calculation.png)

### Multilayer Neural Networks

* While using multilayer neural networks and defining out custom network architecture we can inherit `torch.nn.Module` class which provides many useful functionalities.
For instance, it allows us to encapsulate layers and operations and keep track of the model’s parameters.
* Network structure must be defined in `__init__` constructor of the class.
* We define how the model input and the layers interact with each other in `forward` method. This is called as forward calculation.
* For backward calculations, we use `backward` method of the class but we don't need to implement it as it will automatically be calculated using the loss function and PyTorch's built-in computational graph. So, we wil just call the `backward` function.

In [ ]:
class NeuralNetwork(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs):
        super().__init__()

        self.layers = torch.nn.Sequential(
            # 1st hidden layer
            torch.nn.Linear(num_inputs, 30),  # applies linear or fully connected layer
            torch.nn.ReLU(),  # applies ReLU (rectified linear unit)

            # 2nd hidden layer
            torch.nn.Linear(30, 20),
            torch.nn.ReLU(),

            # output layer
            torch.nn.Linear(20, num_outputs),
        )

    def forward(self, x):
        logits = self.layers(x)  # input X is used in the forward pass calculations of the network
        return logits  # logits is a tensor with the shape of [20, num_outputs]

In [87]:
model = NeuralNetwork(50, 3)

print("Summary of the model structure: ")
print("--------------------------------")
print(model)

Summary of the model structure: 
--------------------------------
NeuralNetwork(
  (layers): Sequential(
    (0): Linear(in_features=50, out_features=30, bias=True)
    (1): ReLU()
    (2): Linear(in_features=30, out_features=20, bias=True)
    (3): ReLU()
    (4): Linear(in_features=20, out_features=3, bias=True)
  )
)


Total number of trainable parameters of the model: 

Each parameter for which `requires_grad=True` counts as a trainable parameter and will be updated during training

In [88]:
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total number of trainable model parameters:", num_params)

Total number of trainable model parameters: 2213


In [91]:
print("Layer's first weight matrix which is the first linear layer:")
print(model.layers[0].weight)
print("")

print("Shape of the linear layer:")
print(model.layers[0].weight.shape)

Layer's first weight matrix which is the first linear layer:
Parameter containing:
tensor([[-0.0230, -0.1265, -0.0904,  ...,  0.0852, -0.0524, -0.0184],
        [ 0.0276, -0.0142, -0.0938,  ..., -0.0250, -0.1225, -0.0009],
        [-0.0710,  0.0062, -0.1141,  ..., -0.0284,  0.0839, -0.1292],
        ...,
        [-0.1318,  0.1222, -0.0349,  ...,  0.1212,  0.0334, -0.0818],
        [ 0.0083,  0.0281,  0.1071,  ..., -0.1123, -0.0380,  0.0435],
        [ 0.0748, -0.0140, -0.0705,  ...,  0.0062,  0.0913,  0.0689]],
       requires_grad=True)

Shape of the linear layer:
torch.Size([30, 50])


The shape of the weight vector of the linear layer is [30, 50] as it is expected.

In [93]:
print("Bias vector:")
print(model.layers[0].bias)
print("")

print("Shape of the bias vector:")
print(model.layers[0].bias.shape)

Bias vector:
Parameter containing:
tensor([-0.0210,  0.0179, -0.0480,  0.0668,  0.0993,  0.1056, -0.1343,  0.1076,
         0.0493, -0.0509, -0.0953, -0.0855, -0.0479,  0.1367, -0.1354, -0.1217,
        -0.1246,  0.1379,  0.0163, -0.0935, -0.0914,  0.1268,  0.0687, -0.0435,
        -0.0843,  0.0109,  0.1317, -0.1144,  0.0789,  0.0469],
       requires_grad=True)

Shape of the bias vector:
torch.Size([30])


The shape of the bias vector of the linear layer is [30] as it is expected.

When we look at the weight and bias vector, we can notice that `requires_grad` is set to True. This assignment is done by default.

We can also notice that the values in weight and bias tensors are small random numbers.
In deep learning, initializing model weights with small random numbers is desired to break symmetry during training. 
Otherwise, the nodes would be performing the same operations and updates during backpropagation, which would not allow the network to learn complex mappings from inputs to outputs.

Nevertheless, we would like to reproduce the weight values in most of the applications during development. Hence, we can easily notice if we made any mistakes. To that attempt, we use `torch.manual_seed` as the following.

In [95]:
torch.manual_seed(123)
model = NeuralNetwork(50, 3)
print(model.layers[0].weight)

Parameter containing:
tensor([[-0.0577,  0.0047, -0.0702,  ...,  0.0222,  0.1260,  0.0865],
        [ 0.0502,  0.0307,  0.0333,  ...,  0.0951,  0.1134, -0.0297],
        [ 0.1077, -0.1108,  0.0122,  ...,  0.0108, -0.1049, -0.1063],
        ...,
        [-0.0787,  0.1259,  0.0803,  ...,  0.1218,  0.1303, -0.1351],
        [ 0.1359,  0.0175, -0.0673,  ...,  0.0674,  0.0676,  0.1058],
        [ 0.0790,  0.1343, -0.0293,  ...,  0.0344, -0.0971, -0.0509]],
       requires_grad=True)


In [97]:
torch.manual_seed(123)
X = torch.rand((1, 50))  # input to the model
out = model(X)  # model automatically runs the forward method and returns the output
print(out)
print("Shape of output: ", out.shape)

tensor([[-0.1262,  0.1080, -0.1792]], grad_fn=<AddmmBackward0>)
Shape of output:  torch.Size([1, 3])


As the model is created with 50 inputs and 3 outputs, the output from the network must have 3 columns. Also, the input to the model has 1 row so the output tensor has a shape of [1, 3].

Here, `grad_fn=<AddmmBackward0>` represents the last-used function to compute a variable in the computational graph. 
In particular, `grad_fn=<AddmmBackward0>` means that the tensor we are inspecting was created via a matrix multiplication and addition operation.
PyTorch will use this information when it computes gradients during backpropagation. 
The `<AddmmBackward0>` part of `grad_fn=<AddmmBackward0>` specifies the operation performed. 
In this case, it is an `Addmm` operation. `Addmm` stands for matrix multiplication (mm) followed by an addition (Add).

Note that if we just want to use a network without training or backpropagation, constructing this computational graph for backpropagation can be wasteful as it performs unnecessary computations and consumes additional memory.

So, when we use a model for inference (for instance, making predictions) rather than training, the best practice is to use the `torch.no_grad()` context manager.

In [98]:
with torch.no_grad():
    out = model(X)

print(out)

tensor([[-0.1262,  0.1080, -0.1792]])


In PyTorch, it’s common practice to code models such that they return the outputs of the last layer (logits) without passing them to a nonlinear activation function.

That’s because PyTorch’s commonly used loss functions combine the softmax (or sigmoid for binary classification) operation with the negative log-likelihood loss in a single class. The reason for this is numerical efficiency and stability. So, if we want to compute class-membership probabilities for our predictions, we have to call the softmax function explicitly.

In [99]:
with torch.no_grad():
    out = torch.softmax(model(X), dim=1)

print(out)

tensor([[0.3113, 0.3934, 0.2952]])


These values can be interpreted as probability values and they sum up to 1.0.

### PyTorch Data Loaders

* PyTorch implements a Dataset and a DataLoader class. 
    * The Dataset class is used to instantiate objects that define how each data record is loaded. 
    * The DataLoader handles how the data is shuffled and assembled into batches.

Example use case:

In [104]:
# Train data
X_train = torch.tensor([
    [-1.4, 3.4],
    [-0.9, 2.9],
    [-0.5, 2.6],
    [2.3, -1.1],
    [2.7, -1.5]
])
y_train = torch.tensor([0, 0, 0, 1, 1])

# Test data
X_test = torch.tensor([
    [-0.8, 2.8],
    [2.6, -1.6],
])
y_test = torch.tensor([0, 1])

display("X_train: ", X_train)
display("y_train: ", y_train)
display("X_test: ", X_test)
display("y_test: ", y_test)

'X_train: '

tensor([[-1.4000,  3.4000],
        [-0.9000,  2.9000],
        [-0.5000,  2.6000],
        [ 2.3000, -1.1000],
        [ 2.7000, -1.5000]])

'y_train: '

tensor([0, 0, 0, 1, 1])

'X_test: '

tensor([[-0.8000,  2.8000],
        [ 2.6000, -1.6000]])

'y_test: '

tensor([0, 1])

Let's create a toy dataset by inheriting PyTorch's Dataset class.

The purpose of this custom ToyDataset class is to instantiate a PyTorch DataLoader.

In [ ]:
from torch.utils.data import Dataset

class ToyDataset(Dataset):
    def __init__(self, X, y):
        self.features = X
        self.labels = y

    def __getitem__(self, index):  # indexing from dataset
        one_x = self.features[index]
        one_y = self.labels[index]
        return one_x, one_y
    
    def __len__(self):  # length of the dataset
        return self.labels.shape[0]
    

train_ds = ToyDataset(X_train, y_train)
test_ds = ToyDataset(X_test, y_test)

Now, we can use PyTorch's DataLoader class to sample from the toy dataset.

In [147]:
from torch.utils.data import DataLoader

train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=0,  # the number of background processes, crucial for parallelizing data loading and preprocessing (0 = main process, not in separate worker process)
    drop_last=True  # when the total number of rows is not evenly divisible by 2, the remaining row will be dropped from the dataset.
)

test_loader = DataLoader(
    dataset=test_ds,
    batch_size=2,
    shuffle=False, # we will not shuffle the test data
    num_workers=0,  # the number of background processes
)

Now, we can iterate over the data loaders

In [148]:
torch.manual_seed(123)  # enables the exact same shuffling order of training examples
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}: ")
    print("----------------")
    display(x)
    display(y)

Batch 1: 
----------------


tensor([[ 2.3000, -1.1000],
        [-0.9000,  2.9000]])

tensor([1, 0])

Batch 2: 
----------------


tensor([[-1.4000,  3.4000],
        [-0.5000,  2.6000]])

tensor([0, 0])

Notes about `num_workers`:

if you are working with tiny datasets or interactive environments such as Jupyter notebooks,
increasing num_workers may not provide any noticeable speedup.
It may, in fact, lead to some problems. One potential problem is the overhead of spinning up multiple worker processes,
which could take longer than the actual data loading when your dataset is small.

Furthermore, for Jupyter notebooks, setting num_workers to greater than 0 can sometimes lead to problems
related to the sharing of resources between different processes, resulting in errors or notebook crashes.

Setting num_workers=4 usually leads to optimal performance on many real-world datasets.

### Training Loop

Let's define a network architecture useful for classification task

In [149]:
class NeuralNetwork(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs):
        super().__init__()

        self.layers = torch.nn.Sequential(
            # 1st hidden layer
            torch.nn.Linear(num_inputs, 30),
            torch.nn.ReLU(),
            # 2nd hidden layer
            torch.nn.Linear(30, 20),
            torch.nn.ReLU(),
            # output layer
            torch.nn.Linear(20, num_outputs),
        )

    def forward(self, x):
        logits = self.layers(x)
        return logits

In [152]:
import torch.nn.functional as F
torch.manual_seed(123)

model = NeuralNetwork(num_inputs=2, num_outputs=2)
optimizer = torch.optim.SGD(
    model.parameters(), lr=0.5  # model parameters are given as input to the optimizer as well as learning rate
)

num_epochs = 3
for epoch in range(num_epochs):
    model.train()  # set the model to training mode

    for batch_idx, (features, labels) in enumerate(train_loader):
        logits = model(features)
        loss = F.cross_entropy(logits, labels)  # cross entropy loss

        optimizer.zero_grad()  # Sets the gradients from the previous round to 0 to prevent unintended gradient accumulation
        loss.backward()  # Computes the gradients
        optimizer.step()  # The optimizer uses the gradients to update the model parameters
        # To prevent undesired gradient accumulation, it is important to include optimizer.zero_grad() call in each update round to reset the gradients to 0.
        # Otherwise, the gradients will accumulate, which may be undesired.

        ### LOGGING
        print(
            f"Epoch: {epoch+1:03d}/{num_epochs:03d}"
            f" | Batch {batch_idx:03d}/{len(train_loader):03d}"
            f" | Train Loss: {loss:.2f}".lstrip()
        )
    
    model.eval()  # set the model to evaluation mode
    # model.eval() is necessary for components that behave differently during training and inference,
    # such as dropout or batch normalization layers. Since we don’t have dropout or other components
    # in our NeuralNetwork class that are affected by these settings, using model.train() and model.eval()
    # is redundant in our preceding code.
    with torch.no_grad():
        outputs = model(X_train)
    print(outputs)

Epoch: 001/003 | Batch 000/002 | Train Loss: 0.75
Epoch: 001/003 | Batch 001/002 | Train Loss: 0.64
tensor([[ 1.1784, -2.1002],
        [ 0.8995, -1.6844],
        [ 0.7006, -1.4043],
        [-0.2737,  0.2677],
        [-0.3151,  0.3243]])
Epoch: 002/003 | Batch 000/002 | Train Loss: 0.44
Epoch: 002/003 | Batch 001/002 | Train Loss: 0.13
tensor([[ 3.1017, -4.5040],
        [ 2.4685, -3.6652],
        [ 2.0365, -3.1067],
        [-1.2497,  1.2496],
        [-1.4464,  1.4626]])
Epoch: 003/003 | Batch 000/002 | Train Loss: 0.03
Epoch: 003/003 | Batch 001/002 | Train Loss: 0.00
tensor([[ 3.2244, -4.6541],
        [ 2.5675, -3.7888],
        [ 2.1203, -3.2121],
        [-1.4791,  1.4793],
        [-1.7154,  1.7320]])


Number of parameters in the network:

In [151]:
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total number of trainable model parameters:", num_params)

Total number of trainable model parameters: 752


In [153]:
torch.set_printoptions(sci_mode=False)  # disables the scientific notation
probas = torch.softmax(outputs, dim=1)
print(probas)

tensor([[    0.9996,     0.0004],
        [    0.9983,     0.0017],
        [    0.9952,     0.0048],
        [    0.0493,     0.9507],
        [    0.0308,     0.9692]])


Let’s consider the first row in the preceding code output. Here, the first value, first row and first column, means that the training example has a 99.96% probability of belonging to class 0 and a 0.04% probability of belonging to class 1.

We can convert these values into class label predictions using PyTorch’s argmax function,
which returns the index position of the highest value in each row if we set dim=1
(setting dim=0 would return the highest value in each column instead):

In [155]:
predictions = torch.argmax(probas, dim=1)

print(predictions)

tensor([0, 0, 0, 1, 1])


<div class="alert alert-block alert-warning">
Note that it is unnecessary to compute softmax probabilities to obtain the class labels.
We could also apply the argmax function to the logits (outputs) directly:
</div>

In [156]:
predictions = torch.argmax(outputs, dim=1)
print(predictions)

tensor([0, 0, 0, 1, 1])


Let's finally check the predictions with target values.

In [158]:
predictions == y_train

tensor([True, True, True, True, True])

Compute the accuracy:

In [164]:
def compute_accuracy(model, dataloader):
    model = model.eval()
    correct = 0.0
    total_examples = 0

    for idx, (features, labels) in enumerate(dataloader):
        with torch.no_grad():
            logits = model(features)
            predictions = torch.argmax(logits, dim=1)
            compare = labels == predictions
            correct += torch.sum(compare)
            total_examples += len(compare)

            return (correct / total_examples).item()  # .item() returns the value of the tensor as a Python float.
        
print("Train accuracy: ", compute_accuracy(model, train_loader))
print("Test accuracy: ", compute_accuracy(model, test_loader))

Train accuracy:  1.0
Test accuracy:  1.0


### Saving and Loading Models

We can save and then load the models to continue training of the model or to utilize the model for testing or inference.

To do so, we can use the following command:

#### Save the model:

In [167]:
# The model’s state_dict is a Python dictionary object that maps each layer in the model to its trainable parameters.
# model.pth is the file path
torch.save(model.state_dict(), "model.pth")

#### Load the model:

In [168]:
model = NeuralNetwork(2, 2)  # create a network class with initial parameters

# torch.load() -> reads the file and reconstructs the model parameters
# torch.load_state_dict() -> applies the read parameters to the model
model.load_state_dict(torch.load("model.pth"))

<All keys matched successfully>

### Training with GPUs

#### Working with a Single GPU

Check if CUDA is available:

In [177]:
print(torch.cuda.is_available())
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

False


If the printed value is False, that means there is not an available CUDA configuration on the system.

In case of available CUDA configuration, you can send the data to CUDA for training. The above `device` has been created in order not to get any error below when there is no CUDA on the device.

In [171]:
# define tensors which will be defined in CPU
tensor_1 = torch.tensor([1., 2., 3.])
tensor_2 = torch.tensor([4., 5., 6.])
print(tensor_1 + tensor_2)

tensor([5., 7., 9.])


In [178]:
tensor_1 = tensor_1.to(device)
tensor_2 = tensor_2.to(device)
print(tensor_1 + tensor_2)

tensor([5., 7., 9.])


We can write the training loop as the following. It will be generic for the systems whether GPU is available on the device or the device is only a CPU device.

In [ ]:
import torch.nn.functional as F
torch.manual_seed(123)

model = NeuralNetwork(num_inputs=2, num_outputs=2).to(device)  # model sent to device (GPU if available else CPU)
optimizer = torch.optim.SGD(
    model.parameters(), lr=0.5  # model parameters are given as input to the optimizer as well as learning rate
)

num_epochs = 3
for epoch in range(num_epochs):
    model.train()  # set the model to training mode

    for batch_idx, (features, labels) in enumerate(train_loader):
        features, labels = features.to(device), labels.to(device)  # features and labels sent to device (GPU if available else CPU)
        logits = model(features)
        loss = F.cross_entropy(logits, labels)  # cross entropy loss

        optimizer.zero_grad()  # Sets the gradients from the previous round to 0 to prevent unintended gradient accumulation
        loss.backward()  # Computes the gradients
        optimizer.step()  # The optimizer uses the gradients to update the model parameters
        # To prevent undesired gradient accumulation, it is important to include optimizer.zero_grad() call in each update round to reset the gradients to 0.
        # Otherwise, the gradients will accumulate, which may be undesired.

        ### LOGGING
        print(
            f"Epoch: {epoch+1:03d}/{num_epochs:03d}"
            f" | Batch {batch_idx:03d}/{len(train_loader):03d}"
            f" | Train Loss: {loss:.2f}".lstrip()
        )
    
    model.eval()  # set the model to evaluation mode
    # model.eval() is necessary for components that behave differently during training and inference,
    # such as dropout or batch normalization layers. Since we don’t have dropout or other components
    # in our NeuralNetwork class that are affected by these settings, using model.train() and model.eval()
    # is redundant in our preceding code.
    with torch.no_grad():
        outputs = model(X_train)
    print(outputs)

### Setting print options:

In [2]:
import torch

b = torch.tensor(
    [[99, 2, 0.6],
     [5, 2, 0.9]]
)

print("PyTorch softmax:")
print(torch.softmax(b, dim=0))

PyTorch softmax:
tensor([[1.0000e+00, 5.0000e-01, 4.2556e-01],
        [1.5008e-41, 5.0000e-01, 5.7444e-01]])


In [3]:
# disable printing as scientific notation
torch.set_printoptions(sci_mode=False)

print(torch.softmax(b, dim=0))

tensor([[    1.0000,     0.5000,     0.4256],
        [    0.0000,     0.5000,     0.5744]])


#### Working with Multiple GPUs

Distributed training is the concept of dividing the model training across multiple GPUs and machines.

You can check Appendix A.9.3 from the book for more detailed information of this section.